From ce96f37a5d82f7033fe80b2bbf9f70a639ef9875 Mon Sep 17 00:00:00 2001
From: Patrik2 Carlsson <patrik2.carlsson@sonymobile.com>
Date: Fri, 10 Jan 2014 15:34:24 +0100
Subject: [PATCH 10/13] Correctly handle unsupported OMX color format

MediaServer was caught in an infinite loop at OMX::getParameter.

OMX::getParameter returns status_t, using OMX_ErrorNoMore to loop
until the index passes beyond the last supported format
will not work since this function only returns ERROR_UNSUPPORTED.

Change-Id: Idde94a46af24641c0e5f1db909335db4c7e2b7f7
---
 media/libstagefright/OMXCodec.cpp | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/media/libstagefright/OMXCodec.cpp b/media/libstagefright/OMXCodec.cpp
index 62bcf3f..f536b95 100644
--- a/media/libstagefright/OMXCodec.cpp
+++ b/media/libstagefright/OMXCodec.cpp
@@ -1348,11 +1348,17 @@ status_t OMXCodec::setVideoOutputFormat(
                 && colorFormat != OMX_COLOR_FormatUnused
                 && colorFormat != format.eColorFormat) {
 
-            while (OMX_ErrorNoMore != err) {
-                format.nIndex++;
+            OMX_U32 index = 1; // Index 0 is retrieved above.
+            while (index < kMaxColorFormatSupported) {
+                format.nIndex = index++;
                 err = mOMX->getParameter(
                         mNode, OMX_IndexParamVideoPortFormat,
                             &format, sizeof(format));
+                if (OK != err) {
+                    format.eColorFormat = OMX_COLOR_FormatUnused;
+                    break;
+                }
+
                 if (format.eColorFormat == colorFormat) {
                     break;
                 }
-- 
2.5.0

